indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java

```diff
@@ -253,46 +287,58 @@ private void requeueBatch(AllocateRequestBatch batch)
-
-    // Process all the batches that are already due
-    int numProcessedBatches = 0;
-    AllocateRequestBatch nextBatch = processingQueue.peekFirst();
-    while (nextBatch != null && nextBatch.isDue()) {
-      // Process the next batch in the queue
-      processingQueue.pollFirst();
-      final AllocateRequestBatch currentBatch = nextBatch;
-      boolean processed;
-      try {
-        processed = processBatch(currentBatch);
-      }
-      catch (Throwable t) {
-        currentBatch.failPendingRequests(t);
-        processed = true;
-        log.error(t, "Error while processing batch[%s].", currentBatch.key);
-      }
-
-      // Requeue if not fully processed yet
-      if (processed) {
-        ++numProcessedBatches;
+    isProcessingScheduled.set(false);
+
+    // Process the batches that are already due
+    int numSubmittedBatches = 0;
+    int numSkippedBatches = 0;
+
+    // Although thread-safe, this iterator might not see entries added by other
+    // concurrent threads. Those entries will be handled in the next processBatchesDue().
+    final Iterator<AllocateRequestBatch> queueIterator = processingQueue.iterator();
+    while (queueIterator.hasNext() && runningDatasources.size() < config.getBatchAllocationNumThreads()) {
+      final AllocateRequestBatch nextBatch = queueIterator.next();
+      final String dataSource = nextBatch.key.dataSource;
+      if (nextBatch.isDue()) {
+        if (runningDatasources.contains(dataSource)) {
+          // Skip this batch as another batch for the same datasource is in progress
+          markBatchAsSkipped(nextBatch);
+          ++numSkippedBatches;
+        } else {
+          // Process this batch
+          queueIterator.remove();
+          runningDatasources.add(dataSource);
+          workerExec.submit(() -> runBatchOnWorker(nextBatch));
+          emitBatchMetric("task/action/batch/submitted", 1L, nextBatch.key);
+          ++numSubmittedBatches;
+        }
       } else {
-        requeueBatch(currentBatch);
+        break;
       }
-
-      nextBatch = processingQueue.peek();
     }
+    log.debug("Submitted [%d] batches, skipped [%d] batches.", numSubmittedBatches, numSkippedBatches);
 
-    // Schedule the next round of processing if the queue is not empty
-    if (processingQueue.isEmpty()) {
-      log.debug("Processed [%d] batches, not scheduling again since queue is empty.", numProcessedBatches);
+    // Schedule the next round of processing if there are available worker threads
+    if (runningDatasources.size() >= config.getBatchAllocationNumThreads()) {
+      log.debug("Not scheduling again since all worker threads are busy.");
+    } else if (numSkippedBatches >= processingQueue.size()) {
+      // All remaining entries in the queue were skipped
+      log.debug("Not scheduling again since datasources are already being processed.");
+    } else if (processingQueue.isEmpty()) {
+      log.debug("Not scheduling again since queue is empty.");
```

Comment:
I think this would be caught by the previous line -- if the queue is empty, `numSkippedBatches` and `processingQueue.size()` are both zero, and `0 >= 0`. Consider collapsing them both into a block like `Not scheduling again since there are no eligible batches (skipped[%d])`

Reply:
Is it okay to keep it as-is for now since it seems to make the compiler happy?

Otherwise it warns about a possible NPE.

<img width="1505" alt="Screenshot 2025-06-11 at 9 11 31 PM" src="https://github.com/user-attachments/assets/393065d6-2150-4a86-8d88-00546c71c445" />

Reply:
Collapsed the condition but still retained the check on `processingQueue.isEmpty()` since it makes the compiler happy.

Otherwise, it warns about a potential NPE.

<img width="1425" alt="Screenshot 2025-06-11 at 9 15 08 PM" src="https://github.com/user-attachments/assets/487f085f-2b2a-4483-a2dc-b326dd809f97" />